File: Source/ChannelStuff/ChannelSprite.cs



using Monocle;
using Microsoft.Xna.Framework;
using Celeste.Mods.auspicioushelper;
using System;
using Celeste.Mod.Entities;

namespace Celeste.Mod.auspicioushelper;

[CustomEntity("auspicioushelper/ChannelSprite")]
public class ChannelSprite:ChannelBaseEntity{
  public int num;
  public Sprite sprite;
  public enum edgeTypes{
    loop, clamp, hide,
  }
  edgeTypes ty; 
  public ChannelSprite(EntityData d, Vector2 offset):base(d.Position+offset){
    channel=d.Attr("channel","");
    Add(sprite=GFX.SpriteBank.Create(d.Attr("xml_spritename")));
    sprite.Position.X=d.Int("offsetX",0);
    sprite.Position.Y=d.Int("offsetY",0);
    if(d.Bool("attached",false)){
      Add(new StaticMover{
        SolidChecker = checkSolid
      });
    }
    num = d.Int("cases",1);
    ty=d.Attr("edge_type","") switch {
      "loop"=>edgeTypes.loop,
      "clamp"=>edgeTypes.clamp,
      _=>edgeTypes.hide,
    };
    Depth=d.Int("depth",2);
  }
  private bool checkSolid(Solid solid){
    return Collide.CheckPoint(solid, Position);
  }
  public override void setChVal(int val){
    if(val<0 || val>=num){
      switch(ty){
        case edgeTypes.loop: val=(val%num+num)%num; break;
        case edgeTypes.clamp: val=Math.Clamp(val,0,num-1);break;
        default:
          sprite.Visible=false;
          return;
      }
    }
    sprite.Visible=true;
    sprite.Play("case"+val.ToString());
  }
  public override void Added(Scene scene){
    base.Added(scene);
    setChVal(ChannelState.readChannel(channel));
  }
}

File: Loenn/entities/channelstuff/channelsprite.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local entity = {}

entity.name = "auspicioushelper/ChannelSprite"
entity.depth = 2000

edge_types = {"loop","clamp","hide"}
entity.placements = {
  {
    name = "Channel Sprite",
    data = {
      channel = "",
      attached = false,
      edge_type = "loop",
      xml_spritename = "auspicioushelper_example1",
      cases=3,
      offsetX=0,
      offsetY=0,
      depth=2
    }
  }
}
entity.fieldInformation = {
  edge_type = {
    options = edge_types
  },
  cases = {
    fieldType="integer"
  },
  offsetX = {
    fieldType="integer"
  },
  offsetY = {
    fieldType="integer"
  },
  depth = {
    fieldType="integer"
  }
}
function entity.sprite(room, entity)
  color = {1, 1, 1, 1}
  local sprite = drawableSprite.fromTexture("util/rect", nil)
  sprite.useRelativeQuad(0, 0, 8, 8) 
  sprite.color = color 
  return sprite
end

function entity.rectangle(room, entity)
  return utils.rectangle(entity.x, entity.y, 8, 8)
end

return entity